File: python/ssync.py
#!/usr/bin/env python
import requests

"""
    # import and init
    import SensorSync from ssync
    s = SensorSync('192.168.2.49')

    # or
    # s = SensorSync()
    # s.ip = '192.168.2.49'


    # set and reset sensors
    # valid sensor numbers: 1..8
    OK, _ = s.activate(1)
    OK, _ = s.reset(1)

    # optionally, check the returned html when activating/resetting:
    OK, _ = s.activate(1, verify=True)
    OK, _ = s.reset(1, verify=True)

    # request status
    OK, status_html = s.status()
 
    # status_html looks like this:
        <html>
        Version 2.3
        IP=192.168.2.49
        MAC=DE.AD.BE.EF.FE.ED
        SENSORS=0
        </html>

    OK, _ = s.write_ip('192.168.2.49')
    OK, _ = s.write_mac('CA.FE.EE.CA.FE.ED')
    OK, _ = s.reboot()
"""

class SensorSync:
    APIS = {
        'activate': 'http://{ip}?activate={sensor}',
        'reset': 'http://{ip}?reset={sensor}',
        'reboot': 'http://{ip}?reboot=true',
        'status': 'http://{ip}?status=true',
        'write_ip': 'http://{ip}?ip={new_ip}',
        'write_mac': 'http://{ip}?mac={new_mac}',
    }

    def __init__(self, ip=None):
        self.ip = ip

    def _url_for(self, api, **kwargs):
        url = SensorSync.APIS[api]
        d = {'ip': self.ip}
        d.update(kwargs)
        return url.format(**d)

    def _request_url(self, url):
        assert self.ip is not None
        r = requests.get(url)
        return r.status_code, r.text

    def activate(self, sensor, verify = False):
        url = self._url_for('activate', sensor=sensor)
        s, c = self._request_url(url)
        OK = s == 200
        if verify:
            expected = '<html>\r\nOK sensor {sensor} activated\r\n</html>\r\n'.format(sensor=sensor)
            OK = OK and c == expected
        return OK, c

    def reset(self, sensor, verify = False):
        url = self._url_for('reset', sensor=sensor)
        s, c = self._request_url(url)
        OK = s == 200
        if verify:
            expected = '<html>\r\nOK sensor {sensor} reset\r\n</html>\r\n'.format(sensor=sensor)
            OK = OK and c == expected
        return OK, c

    def reboot(self, verify):
        url = self._url_for('reboot')
        s, c = self._request_url(url)
        OK = s == 200
        if verify:
            expected = '<html>\r\nOK\r\n</html>\r\n'
            OK = OK and c == expected
        return OK, c

    def status(self):
        url = self._url_for('status')
        return self._request_url(url)

    def write_ip(self, new_ip, verify = False):
        url = self._url_for('write_ip', new_ip=new_ip)
        s, c = self._request_url(url)
        OK = s == 200
        if verify:
            expected = '<html>\r\nOK\r\n</html>\r\n'
            OK = OK and c == expected
        return OK, c

    def write_mac(self, new_mac, verify = False):
        url = self._url_for('write_mac', new_mac=new_mac)
        s, c = self._request_url(url)
        OK = s == 200
        if verify:
            expected = '<html>\r\nOK\r\n</html>\r\n'
            OK = OK and c == expected
        return OK, c



def t(s, t):
    if not s:
        print('*************', end='')
        print(repr(t))
    print(f'status: {s}\n{t}')
    print('---')

if __name__ == '__main__':
    # test against device in default config mode
    s = SensorSync('192.168.2.49')
    t(*s.activate(1, verify=True))
    t(*s.reset(1, verify=True))
    t(*s.status())
    t(*s.write_ip('192.168.2.49', verify=True))
    t(*s.write_mac('CA.FE.EE.CA.FE.ED', verify=True))
    t(*s.reboot(verify=True))

File: javascript/ssync.js
// minimal javascript client

// intended use:
//
//
// set IP address to use:
// ------------------------------------------------------------------------
// set_ssync_ip(ip_address_as_string);
//
//
// activate a sensor:
// ------------------------------------------------------------------------
// activate_sensor(sensor_number);    // sensor number range: 1..8
//
//
// de-activate a sensor:
// ------------------------------------------------------------------------
// reset_sensor(sensor_number);       // sensor number range: 1..8
//
//
// read config:
// ------------------------------------------------------------------------
// read_config()    // returns a promise, access with .then(...)
//
// read_config().then(data => console.log(data.ip, data.mac, data.sensors))
//
//
//
// return value and error handling
// ------------------------------------------------------------------------
//                // just activate
//                activate_sensor(1);
//
//                // activate sensor and log result to console
//                activate_sensor(1).then(result_html => console.log(result_html));
//
//                // activate sensor and log result to console, catch and log error
//                activate_sensor(1)
//                    .then(result_html => console.log(result_html))
//                    .catch(function (err) {
//                        console.log('Fetch Error :-S', err);
//                    });
//
// NOTE: if you chain multiple calls and want them to be executed in order, 
//       use the `await` keyword.  This only works inside of functions:
//
//        async function foo() {
//            console.log('the following happens in order')
//
//            // just activate
//            await activate_sensor(1);
//
//            // activate sensor and log result to console
//            await activate_sensor(1).then(result_html => console.log(result_html));
//
//            // activate sensor and log result to console, catch and log error
//            await activate_sensor(1)
//                .then(result_html => console.log(result_html))
//                .catch(function (err) {
//                    console.log('Fetch Error :-S', err);
//                });
//
//            await read_config().then(data => console.log('\n',
//                'ip: ' + data.ip, '\n',
//                'mac: ' + data.mac, '\n',
//                'sensor states: ', data.sensors, '\n',
//                data)
//            );
//            
//            console.log('finished');
//        }
//
//        // call the function
//        foo();
//
// see `ssync-demo.js` for an example for how to use this inside of an HTML file.


// the default ip
var ssync_ip = '192.168.2.49';

let apis = {
    activate: {
        url: (ip, sensor_number) => {
            return "http://" + ip + "?activate=" + sensor_number
        }
    },
    reset: {
        url: (ip, sensor_number) => {
            return "http://" + ip + "?reset=" + sensor_number
        }
    },
    reboot: {
        url: (ip) => {
            return "http://" + ip + "?reboot=true"
        }
    },
    status: {
        url: (ip) => {
            return "http://" + ip + "?status=true"
        }
    },
    write_ip: {
        url: (ip, new_ip) => {
            return "http://" + ip + "?ip=" + new_ip
        }
    },
    write_mac: {
        url: (ip, new_mac) => {
            return "http://" + ip + "?mac=" + new_mac
        }
    },
};


function set_ssync_ip(ip_address_as_string) {
    ssync_ip = ip_address_as_string;
}


async function activate_sensor(sensor) {
    return fetch(apis.activate.url(ssync_ip, sensor))
        .then(response => response.text())
        .catch(err => {throw err})
}

async function reset_sensor(sensor) {
    return fetch(apis.reset.url(ssync_ip, sensor))
        .then(response => response.text())
        .catch(err => {throw err})
}

async function reboot() {
    return fetch(apis.reboot.url(ssync_ip))
        .then(response => response.text())
        .catch(err => {throw err})
}


function handle_status(data) {
    const lines = data.split('\r\n');

    // we will parse into these variables
    var read_ip = '';
    var read_mac = '';
    var read_sensors = 0;

    for (const line of lines) {
        const keyval = line.split('=');
        if (keyval.length != 2)
            continue;
        const key = keyval[0];
        const value = keyval[1];
        if (key == "MAC") {
            // do something with the mac address
            read_mac = value;
        }
        if (key == "IP") {
            // do something with the mac address
            read_ip = value;
        }
        if (key == "SENSORS") {
            var state = parseInt(value, 16);
            // do something with the states 
            read_sensors = state;
        }
    }
    return {
        'ip': read_ip,
        'mac': read_mac,
        'sensors': read_sensors,
    }
}

async function read_config() {
    return fetch(apis.status.url(ssync_ip))
        .then(response => response.text())
        .then(data => handle_status(data))
        .catch(err => {throw err})
}

async function write_ip(new_ip) {
    return fetch(apis.write_ip.url(ssync_ip, new_ip))
        .then(response => response.text())
        .catch(err => {throw err})
}

async function write_mac(new_mac) {
    return fetch(apis.write_mac.url(new_mac))
        .then(response => response.text())
        .catch(err => {throw err})
}

